jsonalchemy/custom_encoder.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf8 -*-
+#
+# "CustomEncoder" is a derivative of "CustomEncoder" by FastTurtle,
+# licensed under CC-by-SA 3.0 from http://stackoverflow.com/a/17684652/374865
+#
+
+"""Override JSONEncoder's default behavior on dict subclasses."""
+
+from json import JSONEncoder
+from json.encoder import _make_iterencode
+from json.encoder import encode_basestring_ascii
+from json.encoder import encode_basestring
+from json.encoder import FLOAT_REPR
+from json.encoder import INFINITY
+
+class CustomEncoder(JSONEncoder):
+    def iterencode(self, o, _one_shot=False):
+        if self.check_circular:
+            markers = {}
+        else:
+            markers = None
+        if self.ensure_ascii:
+            _encoder = encode_basestring_ascii
+        else:
+            _encoder = encode_basestring
+        if self.encoding != 'utf-8':
+            def _encoder(o, _orig_encoder=_encoder, _encoding=self.encoding):
+                if isinstance(o, str):
+                    o = o.decode(_encoding)
+                return _orig_encoder(o)
+
+        def floatstr(o, allow_nan=self.allow_nan,
+                     _repr=FLOAT_REPR, _inf=INFINITY, _neginf=-INFINITY):
+            if o != o:
+                text = 'NaN'
+            elif o == _inf:
+                text = 'Infinity'
+            elif o == _neginf:
+                text = '-Infinity'
+            else:
+                return _repr(o)
+
+            if not allow_nan:
+                raise ValueError(
+                    "Out of range float values are not JSON compliant: " +
+                    repr(o))
```

Comment:
This class is basically a reimplementation of https://github.com/simplejson/simplejson/blob/master/simplejson/encoder.py#L88, so that custom types that inherit from `dict` or `list` can have custom behavior. This function is similar to https://github.com/simplejson/simplejson/blob/master/simplejson/encoder.py#L301-L326, which hopefully should be tested for us (I noticed that it's not the same, so maybe I'll need to update it).